### ISSUE TITLE
ReplicationControllerManager is too slow to support 1000-node clusters

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#21672]
I've finally debugged the main problem that is causing load.go test failures in 1000-node clusters.

The problem is basically getPodController() method:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L202

Here is exactly what is happening:
- in load test in 1000-node cluster, we create ~3500 replication controllers, which altogether have 30.000 pods
- then for each pod ReplicationControllerManager is creating a batch of pods:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L344
- and then is waiting for them to be observed via watch
- however, on top of watch, we have an Informer framework in ReplicationControllerManager:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L152
- and the problem is that a single addPod/updatePod/deletePod call (which are called sequentially for every single add/update/delete pod action) is calling: getPodController:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L202
- and getPodController has complexity O(#replicationControllers)

Which means that for any pod add/update/delete done in the system, ReplicationControllerManager is doing O(#replicationControllers) computations (which are btw non-trivial ones).

This is basically, why density.go (where we have a single RC) works pretty good, but load.go test basically doesn't work at all.

In my opinion this is a blocker for 1.2.

@fgrzadkowski @gmarek @davidopp @lavalamp @kubernetes/goog-control-plane @kubernetes/sig-scalability

**Comment by wojtek-t** [comment]
I think that what we should do for 1.2 is some kind of parallelization.
The solution that I would suggest is:
- in the Informer framework add a support for processing events in parallel
- however, just to avoid problems, we should ensure that a given object would always be processed by exactly the same worker (that way we e.g. avoid problems of processing the same object syncurrently multiple times)
- we can do it by creating multiple workers, and each worker processing objects with UID modulo #workers = worker number

I think that processing multiple pods in parallel shouldn't break anything in replication controller.

Thoughts? @lavalamp @davidopp

**Comment by wojtek-t** [comment]
Although - to be clear, what I'm suggesting above is a short-term fix for 1.2 - we should think about some long-term solution too (e.g. introducing Indexing or sth like that, but that shouldn't be a blocker for 1.2).

**Comment by wojtek-t** [comment]
https://github.com/kubernetes/kubernetes/pull/21674 is showing how I would imagine parallelization that I described above (this is just proof-of-concept). Any comments are welcome.

**Comment by AdoHe** [comment]
:+1:

**Comment by AdoHe** [comment]
+1

**Comment by quinton-hoole** [comment]
@wojtek-t Awesome debugging!  
Your proposed solution seems reasonable, although can't you avoid processing the same object concurrently multiple times with a relatively simple channel?  That way you could avoid the sharding complexities you mention (even though the complexity is not that high).

**Comment by bprashanth** [comment]
We've come a long way if in memory computations are now the bottleneck for the replication controller manager :)

https://github.com/kubernetes/kubernetes/pull/19043 is an attempt to serve the results of the store list from a lookup cache

**Comment by bprashanth** [comment]
> and the problem is that a single addPod/updatePod/deletePod call (which are called sequentially for every single add/update/delete pod action) is calling: getPodController:

Sequentially within a single worker. The RC manager runs 5 workers by default, though, and they should process RCs in parallel: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L184

**Comment by lavalamp** [comment]
I think I have a PR addressing this sitting in my review queue for weeks now. I'll look for it shortly.

**Comment by bprashanth** [comment]
I think that's https://github.com/kubernetes/kubernetes/pull/19043

**Comment by wojtek-t** [comment]
>
Sequentially within a single worker. The RC manager runs 5 workers by default, though, and they should process RCs in parallel: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L184

This is not a problem - the problem is processing pods within a controller framework which is done sequentially.
RCs are processed by workers, 5 in parallel, but this is not a bottleneck.

**Comment by wojtek-t** [comment]
Yeah - it seems that #19043 would also address this issue.
However, I'm not convinved about that PR - I will comment on it.

**Comment by bgrant0607** [comment]
cc @mqliang

**Comment by bgrant0607** [comment]
If you make any changes to ReplicationController, please make the same changes to ReplicaSet.

cc @madhusudancs

**Comment by bgrant0607** [comment]
controllerRef (#14961) could effectively act as an index. We should do that in 1.3.

**Comment by bgrant0607** [comment]
In the mean time, it seems like it would be pretty easy to memoize the result of the lookup, recheck that the selector matches, invalidate the cached result if it didn't still match, and use the cached result otherwise.

**Comment by mikedanese** [comment]
DaemonSetController likely has similar issues

**Comment by wojtek-t** [comment]
@bgrant0607 - it's easy if we assume that we don't have overlapping sectors. If we don't (which is what the code is currently doing), creating a new RC can result in changing belonging of different pods (this is what is currently happening - in other words creating an RC can effectively invalidate all existing pods in the worst case - or at least we should iterate over all pods).
It's definitely possible to solve it like that, but what I'm proposing in #21674 seems to be enough for 1.2 and less controversial.

**Comment by wojtek-t** [comment]
>
DaemonSetController likely has similar issues

@mikedanese - yeah, but to be honest it seems to be that there would be only a couple of DaemonSets in a single cluster (like O(10)). If that's the case, it shouldn't be the problem now ;)
But thanks for pointing on it - I will keep that in mind.

**Comment by bgrant0607** [comment]
@wojtek-t We break overlap ties using creation timestamp. If an older RC still exists, a new one can't take over its pods, so creating a new RC shouldn't invalidate existing memoized results.

**Comment by bgrant0607** [comment]
I don't object to parallelization. The 2 approaches can be complementary.

**Comment by bprashanth** [comment]
> @wojtek-t We break overlap ties using creation timestamp. If an older RC still exists, a new one can't take over its pods, so creating a new RC shouldn't invalidate existing memoized results.

Sorting by timestamp is a sub-optimal solution because an single RC can manage pod with different labels. So if we have 3 RCs a,b,c, and podA matches a,b, and we choose a from timestamp, then podB comes in and matches a,b,c and we choose c by timestamp, we're back to fighting. 

This is just another case of overlap we fumble.

**Comment by wojtek-t** [comment]
@bgrant0607 - ok thanks. To be clear, I also don't object caching (in fact I think we need it, but we may need to be careful with it).

**Comment by lavalamp** [comment]
I also vote for memoization. The more I think about it, the more I think adding parellization into the controller framework/informer is very dangerous. I don't think consumers of that code are expecting their handlers to be called in parallel right now.

**Comment by bprashanth** [comment]
>I also vote for memoization. The more I think about it, the more I think adding parellization into the controller framework/informer is very dangerous. I don't think consumers of that code are expecting their handlers to be called in parallel right now.

+1. Many controllers treat the serialization in the Informer like a write lock, including the Ingress controllers in contrib and who-knows-what downstream.

**Comment by lavalamp** [comment]
Also take a look at #17277.

**Comment by mqliang** [comment]
> Also take a look at #17277.

Yes, I have two PRs which try to speed up `getPodController()` method:
1) https://github.com/kubernetes/kubernetes/pull/17277, which add namespace index  for cache
2) https://github.com/kubernetes/kubernetes/pull/19043, which add a look-up cache

I will rebase them soon if the optimization skills looks good. 

And there is a issue which point out a another potential problem of `getPodController()` for a larger cluster: https://github.com/kubernetes/kubernetes/issues/21127

**Comment by wojtek-t** [comment]
@mqliang - #19043 is much more important for now. So please focus on it first. I will try to look into it today too.

**Comment by wojtek-t** [comment]
@lavalamp @bprashanth - I agree that we should NOT change the controller framework to be non-serial in general. But what I'm suggesting is to add a possiblity to process coming events in parallel.
So if we set the number of workers to 1, we are back to processing the in serial in the same order in which they are coming. We can make this a default behavior, so it will be totally up to users to switch it on if they are comfortable with it. Note that this will work fine in case of ReplicationController.

**Comment by mqliang** [comment]
> #19043 is much more important for now. So please focus on it first. I will try to look into it today too.
@wojtek-t OK, I will rebase it soon and apply the same optimization skill to ReplicaSet(And maybe DaemonSet)

**Comment by wojtek-t** [comment]
@mqliang - thanks!
Please apply it to ReplicaSet - we can change DaemonSet later (or at least in separate PR).

**Comment by wojtek-t** [comment]
@bgrant0607 @lavalamp 

>
@wojtek-t We break overlap ties using creation timestamp. If an older RC still exists, a new one can't take over its pods, so creating a new RC shouldn't invalidate existing memoized results.

Yes, but the problem is if a pod is attached to some new RC, and then you change the selector of older RC so that it is now matching that pod. Then if we use caching, we won't observe the change (which we would if we were computing it from scratch every time) - see my comment:
https://github.com/kubernetes/kubernetes/pull/19043/files#r53750770

**Comment by mqliang** [comment]
BTW, usually, a rc will have a replica number 3-5, so https://github.com/kubernetes/kubernetes/pull/19043 could only speed up `getPodController()` about 3-5 times. 

If https://github.com/kubernetes/kubernetes/pull/19043 is not enoght to fix this issue, we may need https://github.com/kubernetes/kubernetes/pull/17277, too. https://github.com/kubernetes/kubernetes/pull/17277 will have a 2 times improvement in case of 100 namespaces each with 10 RCs each with 10 pods, and 20 times improvement in case of 100 namespaces each with 10 RCs each with 10 pods.

**Comment by bgrant0607** [comment]
@wojtek-t Yes, good point. A selector change is another case where the cached backrefs would need to be invalidated. An alternative is that we could implement controllerRef, which wouldn't require invalidation in that case.

**Comment by bgrant0607** [comment]
@mqliang Yes, many users will use small RCs. Some will use fairly large ones, however, and we do want to scale in both scenarios.

**Comment by wojtek-t** [comment]
@bgrant0607 - since this is too close to 1.2 to do anything very sophisticated, I think that what @mqliang suggested which is invalidating the whole cache in case of update of any ReplicationController that is changing the selector is good enough.
I think that changing selector of ReplicationController is rare enough so that we can leave with it in 1.2 and come up with something better for 1.3 when we will have much more time to think about it. WDYT?

**Comment by bgrant0607** [comment]
I agree.

**Comment by wojtek-t** [comment]
With #19043 merge we had a first successful run of load test in 1000-node kubemark. If this repeats, I will close this issue - @davidopp - FYI

**Comment by wojtek-t** [comment]
Yes - it seems that this issue is solved
